openshift/WFLY-15421_OIDC_for_WildFly_s2i_v2.adoc

```diff
@@ -0,0 +1,306 @@
+
+This implies a set of WildFly CLI management operations to be called at packaging time.
+
+For example:
+`/subsystem=elytron-oidc-client/secure-deployment=simple-webapp.war:add(client-id=simple-webapp, public-client=true, provider-url=${OIDC_URL}, ssl-required=EXTERNAL)`
+
+NB: This way of configuring a deployment implies that a “client” has been created in the authorization server. That is expected.
+
+==== Environment variable based configuration
+
+This is the main topic of this proposal.
+
+A CLI script is generated at boot time to handle the OIDC subsystem configuration based on discovered 
+deployments (actually deployed and found in content repository or located in `standalone/deployments` directory).
+In addition there is an automatic creation and registration of the clients in the authorization server.
+
+In order for war deployments to be discovered they must have the `OIDC` auth-method in web.xml. 
+Deployments that contain a `WEB-INF/oidc.json` descriptor file are ignored.
+
+NB: A single provider can be configured for a running server. The discovered deployments are all registered inside the same OIDC provider.
+
+We are offering a new set of env variables `OIDC_*` to be used to configure the subsystem, enable the discovery of deployments and registration of 
+clients into the authorization server.
+
+The existing `SSO_*` env variables used for OIDC are usable for Keycloak OIDC integration. 
+It helps transition smoothly to use the native OIDC support. 
+This proposal defines a mapping between the `SSO_*` and `OIDC_*` env variables.
+The `SSO_*` env variables are deprecated.
+
+The env variable `OIDC_PROVIDER_NAME=<name>` MUST be set when using `OIDC_*` env variables. When using `SSO_*` env variables 
+the provider name `keycloak` is internally used.
+
+A mix of `SSO_*` and `OIDC_*` env variables, although not recommended, would work, 
+new env variables will only be added in the `OIDC_*` namespace. We have a first example `OIDC_PROVIDER_SSL_REQUIRED` 
+that would be new without a mapping in the `SSO_*` namespace.
+
+Example of configuration using new OIDC env variables for Keycloak:
+
+```
+OIDC_PROVIDER_NAME=keycloak
+OIDC_PROVIDER_URL=http://keycloak/auth/realms/Wildfly
+OIDC_USER_NAME=foo
+OIDC_USER_PASSWORD=foo
+OIDC_HOSTNAME_HTTP=my-app-myproject.192.168.42.233.nip.io
+```
+
+Example of configuration using the existing SSO env variables:
+
+```
+SSO_URL=http://keycloak/auth
+SSO_REALM=Wildfly
+SSO_USER=foo
+SSO_PASSWORD=foo
+HOSTNAME_HTTP=my-app-myproject.192.168.42.233.nip.io
+```
+
+Example of configuration using new OIDC env variables for a custom provider:
+
+```
+OIDC_PROVIDER_NAME=other
+OIDC_PROVIDER_URL=http://foo/specific/url/for/provider
+OIDC_HOSTNAME_HTTP=my-app-myproject.192.168.42.233.nip.io
+```
+
+===== Env variables mapping table 
+
+|===
+|Env var |SSO|Details|Required|Defaults
```

Comment:
I think new set of variable with prefix OIDC just make configuration more complex. Is not enough add variable for OIDC_PROVIDER_SSL_REQUIRED with prefix SSO instead OIDC?

Reply:
The SSO_XXX are kept for backward compatibility with Keycloak/SSO adapters. The OIDC_XXX are introduced as new env variable that are in use with the elytron-oidc-client. elytron-oidc-client is an open adapter that allows to connect to other oidc provider than RH SSO/Keycloak.

Reply:
I wanted to point out the SSS_XXX and OIDC_XXX variables are almost same. So the elytron-oidc-client could be configured by SSO_XXX varibles. then the OIDC_PROVIDER_NAME should have some specific value to use Keycloak adapter and different one for elytron-oidc-client. For example as it is used in web.xml. auth-method must be OIDC to use elytron-oidc-client and KEYCLOAK for Keycloak adapter.

Reply:
resolved in https://github.com/wildfly/wildfly-proposals/pull/434#discussion_r970853665